.local/perl/man/man3/strictures.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "strictures 3"
.TH strictures 3 "2014-08-16" "perl v5.20.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
strictures \- turn on strict and make all warnings fatal
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  use strictures 1;
.Ve
.PP
is equivalent to
.PP
.Vb 2
\&  use strict;
\&  use warnings FATAL => \*(Aqall\*(Aq;
.Ve
.PP
except when called from a file which matches:
.PP
.Vb 1
\&  (caller)[1] =~ /^(?:t|xt|lib|blib)/
.Ve
.PP
and when either \f(CW\*(C`.git\*(C'\fR, \f(CW\*(C`.svn\*(C'\fR, or \f(CW\*(C`.hg\*(C'\fR is present in the current directory (with
the intention of only forcing extra tests on the author side) \*(-- or when \f(CW\*(C`.git\*(C'\fR,
\&\f(CW\*(C`.svn\*(C'\fR, or \f(CW\*(C`.hg\*(C'\fR is present two directories up along with \f(CW\*(C`dist.ini\*(C'\fR (which would
indicate we are in a \f(CW\*(C`dzil test\*(C'\fR operation, via Dist::Zilla) \*(--
or when the \f(CW\*(C`PERL_STRICTURES_EXTRA\*(C'\fR environment variable is set, in which case
.PP
.Vb 1
\&  use strictures 1;
.Ve
.PP
is equivalent to
.PP
.Vb 5
\&  use strict;
\&  use warnings FATAL => \*(Aqall\*(Aq;
\&  no indirect \*(Aqfatal\*(Aq;
\&  no multidimensional;
\&  no bareword::filehandles;
.Ve
.PP
Note that \f(CW\*(C`PERL_STRICTURES_EXTRA\*(C'\fR may at some point add even more tests, with only a minor
version increase, but any changes to the effect of \f(CW\*(C`use strictures\*(C'\fR in
normal mode will involve a major version bump.
.PP
If any of the extra testing modules are not present, strictures will
complain loudly, once, via \f(CW\*(C`warn()\*(C'\fR, and then shut up. But you really
should consider installing them, they're all great anti-footgun tools.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
I've been writing the equivalent of this module at the top of my code for
about a year now. I figured it was time to make it shorter.
.PP
Things like the importer in \f(CW\*(C`use Moose\*(C'\fR don't help me because they turn
warnings on but don't make them fatal \*(-- which from my point of view is
useless because I want an exception to tell me my code isn't warnings-clean.
.PP
Any time I see a warning from my code, that indicates a mistake.
.PP
Any time my code encounters a mistake, I want a crash \*(-- not spew to \s-1STDERR\s0
and then unknown (and probably undesired) subsequent behaviour.
.PP
I also want to ensure that obvious coding mistakes, like indirect object
syntax (and not so obvious mistakes that cause things to accidentally compile
as such) get caught, but not at the cost of an \s-1XS\s0 dependency and not at the
cost of blowing things up on another machine.
.PP
Therefore, strictures turns on additional checking, but only when it thinks
it's running in a test file in a \s-1VCS\s0 checkout \*(-- although if this causes
undesired behaviour this can be overridden by setting the
\&\f(CW\*(C`PERL_STRICTURES_EXTRA\*(C'\fR environment variable.
.PP
If additional useful author side checks come to mind, I'll add them to the
\&\f(CW\*(C`PERL_STRICTURES_EXTRA\*(C'\fR code path only \*(-- this will result in a minor version increase (e.g.
1.000000 to 1.001000 (1.1.0) or similar). Any fixes only to the mechanism of
this code will result in a sub-version increase (e.g. 1.000000 to 1.000001
(1.0.1)).
.PP
If the behaviour of \f(CW\*(C`use strictures\*(C'\fR in normal mode changes in any way, that
will constitute a major version increase \*(-- and the code already checks
when its version is tested to ensure that
.PP
.Vb 1
\&  use strictures 1;
.Ve
.PP
will continue to only introduce the current set of strictures even if 2.0 is
installed.
.SH "METHODS"
.IX Header "METHODS"
.SS "import"
.IX Subsection "import"
This method does the setup work described above in \*(L"\s-1DESCRIPTION\*(R"\s0
.SS "\s-1VERSION\s0"
.IX Subsection "VERSION"
This method traps the \f(CW\*(C`strictures\->VERSION(1)\*(C'\fR call produced by a use line
with a version number on it and does the version check.
.SH "EXTRA TESTING RATIONALE"
.IX Header "EXTRA TESTING RATIONALE"
Every so often, somebody complains that they're deploying via \f(CW\*(C`git pull\*(C'\fR
and that they don't want strictures to enable itself in this case \*(-- and that
setting \f(CW\*(C`PERL_STRICTURES_EXTRA\*(C'\fR to 0 isn't acceptable (additional ways to
disable extra testing would be welcome but the discussion never seems to get
that far).
.PP
In order to allow us to skip a couple of stages and get straight to a
productive conversation, here's my current rationale for turning the
extra testing on via a heuristic:
.PP
The extra testing is all stuff that only ever blows up at compile time;
this is intentional. So the oft-raised concern that it's different code being
tested is only sort of the case \*(-- none of the modules involved affect the
final optree to my knowledge, so the author gets some additional compile
time crashes which he/she then fixes, and the rest of the testing is
completely valid for all environments.
.PP
The point of the extra testing \*(-- especially \f(CW\*(C`no indirect\*(C'\fR \*(-- is to catch
mistakes that newbie users won't even realise are mistakes without
help. For example,
.PP
.Vb 1
\&  foo { ... };
.Ve
.PP
where foo is an & prototyped sub that you forgot to import \*(-- this is
pernicious to track down since all \fIseems\fR fine until it gets called
and you get a crash. Worse still, you can fail to have imported it due
to a circular require, at which point you have a load order dependent
bug which I've seen before now \fIonly\fR show up in production due to tiny
differences between the production and the development environment. I wrote
<http://shadow.cat/blog/matt\-s\-trout/indirect\-but\-still\-fatal/> to explain
this particular problem before strictures itself existed.
.PP
As such, in my experience so far strictures' extra testing has
\&\fIavoided\fR production versus development differences, not caused them.
.PP
Additionally, strictures' policy is very much \*(L"try and provide as much
protection as possible for newbies \*(-- who won't think about whether there's
an option to turn on or not\*(R" \*(-- so having only the environment variable
is not sufficient to achieve that (I get to explain that you need to add
\&\f(CW\*(C`use strict\*(C'\fR at least once a week on freenode #perl \*(-- newbies sometimes
completely skip steps because they don't understand that that step
is important).
.PP
I make no claims that the heuristic is perfect \*(-- it's already been evolved
significantly over time, especially for 1.004 where we changed things to
ensure it only fires on files in your checkout (rather than strictures\-using
modules you happened to have installed, which was just silly). However, I
hope the above clarifies why a heuristic approach is not only necessary but
desirable from a point of view of providing new users with as much safety as possible,
and will allow any future discussion on the subject to focus on \*(L"how do we
minimise annoyance to people deploying from checkouts intentionally\*(R".
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.IP "\(bu" 4
indirect
.IP "\(bu" 4
multidimensional
.IP "\(bu" 4
bareword::filehandles
.SH "COMMUNITY AND SUPPORT"
.IX Header "COMMUNITY AND SUPPORT"
.SS "\s-1IRC\s0 channel"
.IX Subsection "IRC channel"
irc.perl.org #toolchain
.PP
(or bug 'mst' in query on there or freenode)
.SS "Git repository"
.IX Subsection "Git repository"
Gitweb is on http://git.shadowcat.co.uk/ and the clone \s-1URL\s0 is:
.PP
.Vb 1
\&  git clone git://git.shadowcat.co.uk/p5sagit/strictures.git
.Ve
.PP
The web interface to the repository is at:
.PP
.Vb 1
\&  http://git.shadowcat.co.uk/gitweb/gitweb.cgi?p=p5sagit/strictures.git
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
mst \- Matt S. Trout (cpan:MSTROUT) <mst@shadowcat.co.uk>
.SH "CONTRIBUTORS"
.IX Header "CONTRIBUTORS"
Karen Etheridge (cpan:ETHER) <ether@cpan.org>
.PP
Mithaldu \- Christian Walde (cpan:MITHALDU) <walde.christian@gmail.com>
.PP
haarg \- Graham Knop (cpan:HAARG) <haarg@haarg.org>
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2010 the strictures \*(L"\s-1AUTHOR\*(R"\s0 and \*(L"\s-1CONTRIBUTORS\*(R"\s0
as listed above.
.SH "LICENSE"
.IX Header "LICENSE"
This library is free software and may be distributed under the same terms
as perl itself.


.cpanm/work/1411440397.17004/strictures-1.005005/blib/lib/strictures.pm
package strictures;

use strict;
use warnings FATAL => 'all';

BEGIN {
  *_PERL_LT_5_8_4 = ($] < 5.008004) ? sub(){1} : sub(){0};
}

our $VERSION = '1.005005';

sub VERSION {
  my ($class, $version) = @_;
  for ($version) {
    last unless defined && !ref && int != 1;
    die "Major version specified as $_ - this is strictures version 1";
  }
  # passing undef here may either warn or die depending on the version of perl.
  # we can't match the caller's warning state in this case, so just disable the
  # warning.
  no warnings 'uninitialized';
  shift->SUPER::VERSION(@_);
}

our $extra_load_states;

our $Smells_Like_VCS = (-e '.git' || -e '.svn' || -e '.hg'
  || (-e '../../dist.ini'
      && (-e '../../.git' || -e '../../.svn' || -e '../../.hg' )));

sub import {
  strict->import;
  warnings->import(FATAL => 'all');

  my $extra_tests = do {
    if (exists $ENV{PERL_STRICTURES_EXTRA}) {
      if (_PERL_LT_5_8_4 and $ENV{PERL_STRICTURES_EXTRA}) {
        die 'PERL_STRICTURES_EXTRA checks are not available on perls older than 5.8.4: '
          . "please unset \$ENV{PERL_STRICTURES_EXTRA}\n";
      }
      $ENV{PERL_STRICTURES_EXTRA};
    } elsif (! _PERL_LT_5_8_4) {
      !!((caller)[1] =~ /^(?:t|xt|lib|blib)/
         and $Smells_Like_VCS)
    }
  };
  if ($extra_tests) {
    $extra_load_states ||= do {

      my (%rv, @failed);
      foreach my $mod (qw(indirect multidimensional bareword::filehandles)) {
        eval "require $mod; \$rv{'$mod'} = 1;" or do {
          push @failed, $mod;

          # courtesy of the 5.8 require bug
          # (we do a copy because 5.16.2 at least uses the same read-only
          # scalars for the qw() list and it doesn't seem worth a $^V check)

          (my $file = $mod) =~ s|::|/|g;
          delete $INC{"${file}.pm"};
        };
      }

      if (@failed) {
        my $failed = join ' ', @failed;
        print STDERR <<EOE;
strictures.pm extra testing active but couldn't load all modules. Missing were:

  $failed

Extra testing is auto-enabled in checkouts only, so if you're the author
of a strictures-using module you need to run:

  cpan indirect multidimensional bareword::filehandles

but these modules are not required by your users.
EOE
      }

      \%rv;
    };

    indirect->unimport(':fatal') if $extra_load_states->{indirect};
    multidimensional->unimport if $extra_load_states->{multidimensional};
    bareword::filehandles->unimport if $extra_load_states->{'bareword::filehandles'};
  }
}

1;

__END__
=head1 NAME

strictures - turn on strict and make all warnings fatal

=head1 SYNOPSIS

  use strictures 1;

is equivalent to

  use strict;
  use warnings FATAL => 'all';

except when called from a file which matches:

  (caller)[1] =~ /^(?:t|xt|lib|blib)/

and when either C<.git>, C<.svn>, or C<.hg> is present in the current directory (with
the intention of only forcing extra tests on the author side) -- or when C<.git>,
C<.svn>, or C<.hg> is present two directories up along with C<dist.ini> (which would
indicate we are in a C<dzil test> operation, via L<Dist::Zilla>) --
or when the C<PERL_STRICTURES_EXTRA> environment variable is set, in which case

  use strictures 1;

is equivalent to

  use strict;
  use warnings FATAL => 'all';
  no indirect 'fatal';
  no multidimensional;
  no bareword::filehandles;

Note that C<PERL_STRICTURES_EXTRA> may at some point add even more tests, with only a minor
version increase, but any changes to the effect of C<use strictures> in
normal mode will involve a major version bump.

If any of the extra testing modules are not present, L<strictures> will
complain loudly, once, via C<warn()>, and then shut up. But you really
should consider installing them, they're all great anti-footgun tools.

=head1 DESCRIPTION

I've been writing the equivalent of this module at the top of my code for
about a year now. I figured it was time to make it shorter.

Things like the importer in C<use Moose> don't help me because they turn
warnings on but don't make them fatal -- which from my point of view is
useless because I want an exception to tell me my code isn't warnings-clean.

Any time I see a warning from my code, that indicates a mistake.

Any time my code encounters a mistake, I want a crash -- not spew to STDERR
and then unknown (and probably undesired) subsequent behaviour.

I also want to ensure that obvious coding mistakes, like indirect object
syntax (and not so obvious mistakes that cause things to accidentally compile
as such) get caught, but not at the cost of an XS dependency and not at the
cost of blowing things up on another machine.

Therefore, L<strictures> turns on additional checking, but only when it thinks
it's running in a test file in a VCS checkout -- although if this causes
undesired behaviour this can be overridden by setting the
C<PERL_STRICTURES_EXTRA> environment variable.

If additional useful author side checks come to mind, I'll add them to the
C<PERL_STRICTURES_EXTRA> code path only -- this will result in a minor version increase (e.g.
1.000000 to 1.001000 (1.1.0) or similar). Any fixes only to the mechanism of
this code will result in a sub-version increase (e.g. 1.000000 to 1.000001
(1.0.1)).

If the behaviour of C<use strictures> in normal mode changes in any way, that
will constitute a major version increase -- and the code already checks
when its version is tested to ensure that

  use strictures 1;

will continue to only introduce the current set of strictures even if 2.0 is
installed.

=head1 METHODS

=head2 import

This method does the setup work described above in L</DESCRIPTION>

=head2 VERSION

This method traps the C<< strictures->VERSION(1) >> call produced by a use line
with a version number on it and does the version check.

=head1 EXTRA TESTING RATIONALE

Every so often, somebody complains that they're deploying via C<git pull>
and that they don't want L<strictures> to enable itself in this case -- and that
setting C<PERL_STRICTURES_EXTRA> to 0 isn't acceptable (additional ways to
disable extra testing would be welcome but the discussion never seems to get
that far).

In order to allow us to skip a couple of stages and get straight to a
productive conversation, here's my current rationale for turning the
extra testing on via a heuristic:

The extra testing is all stuff that only ever blows up at compile time;
this is intentional. So the oft-raised concern that it's different code being
tested is only sort of the case -- none of the modules involved affect the
final optree to my knowledge, so the author gets some additional compile
time crashes which he/she then fixes, and the rest of the testing is
completely valid for all environments.

The point of the extra testing -- especially C<no indirect> -- is to catch
mistakes that newbie users won't even realise are mistakes without
help. For example,

  foo { ... };

where foo is an & prototyped sub that you forgot to import -- this is
pernicious to track down since all I<seems> fine until it gets called
and you get a crash. Worse still, you can fail to have imported it due
to a circular require, at which point you have a load order dependent
bug which I've seen before now I<only> show up in production due to tiny
differences between the production and the development environment. I wrote
L<http://shadow.cat/blog/matt-s-trout/indirect-but-still-fatal/> to explain
this particular problem before L<strictures> itself existed.

As such, in my experience so far L<strictures>' extra testing has
I<avoided> production versus development differences, not caused them.

Additionally, L<strictures>' policy is very much "try and provide as much
protection as possible for newbies -- who won't think about whether there's
an option to turn on or not" -- so having only the environment variable
is not sufficient to achieve that (I get to explain that you need to add
C<use strict> at least once a week on freenode #perl -- newbies sometimes
completely skip steps because they don't understand that that step
is important).

I make no claims that the heuristic is perfect -- it's already been evolved
significantly over time, especially for 1.004 where we changed things to
ensure it only fires on files in your checkout (rather than L<strictures>-using
modules you happened to have installed, which was just silly). However, I
hope the above clarifies why a heuristic approach is not only necessary but
desirable from a point of view of providing new users with as much safety as possible,
and will allow any future discussion on the subject to focus on "how do we
minimise annoyance to people deploying from checkouts intentionally".

=head1 SEE ALSO

=over 4

=item *

L<indirect>

=item *

L<multidimensional>

=item *

L<bareword::filehandles>

=back

=head1 COMMUNITY AND SUPPORT

=head2 IRC channel

irc.perl.org #toolchain

(or bug 'mst' in query on there or freenode)

=head2 Git repository

Gitweb is on http://git.shadowcat.co.uk/ and the clone URL is:

  git clone git://git.shadowcat.co.uk/p5sagit/strictures.git

The web interface to the repository is at:

  http://git.shadowcat.co.uk/gitweb/gitweb.cgi?p=p5sagit/strictures.git

=head1 AUTHOR

mst - Matt S. Trout (cpan:MSTROUT) <mst@shadowcat.co.uk>

=head1 CONTRIBUTORS

Karen Etheridge (cpan:ETHER) <ether@cpan.org>

Mithaldu - Christian Walde (cpan:MITHALDU) <walde.christian@gmail.com>

haarg - Graham Knop (cpan:HAARG) <haarg@haarg.org>

=head1 COPYRIGHT

Copyright (c) 2010 the strictures L</AUTHOR> and L</CONTRIBUTORS>
as listed above.

=head1 LICENSE

This library is free software and may be distributed under the same terms
as perl itself.

=cut
